### ISSUE TITLE
Releases

### ISSUE THREAD

**Issue by UebelAndre** [bazelbuild/rules_rust#415]
Why aren't there releases for this repo? I feel like this is something every rules repository should have. Is there anything stopping this repo from having releases?

**Comment by damienmg** [comment]
We are all working in companies that use a work at head model in which release does not really make sense. As this repositories is supposed to be used at source level there is no compelling reason to have release and release induce toil.

**Comment by mfarrugi** [comment]
I think that it is useful to tag the minimum (and maximum?) bazel version
required, since those are the only breaking changes we have made in the
past. This was a source of toil for me in the past, lining things up w/ the
very frequent bazel updates.

I think the 'live at head' approach has only been the case for
contributors, as most people will want to include their changes by using
the latest version.

**Comment by UebelAndre** [comment]
> I think that it is useful to tag the minimum (and maximum?) bazel version required

> most people will want to include their changes by using the latest version.

I agree with both of these points.

**Comment by UebelAndre** [comment]
Would it be possible to get a release? Is there something in CI that tests the oldest supported version?

**Comment by mfarrugi** [comment]
Sure, but what do you need from a release? It is little more than a tag.

It looks like CI can (
https://github.com/bazelbuild/continuous-integration/blob/master/buildkite/README.md#using-a-specific-version-of-bazel),
but currently doesn't.

**Comment by UebelAndre** [comment]
A release to me is an indication of what kinds of changes were made to a code base. It's little more than a tag but the value in it is the semantic versioning. If a neat new feature was added, I should be able to infer that from the release number.

I think testing should be done on some modern version and it should be up to the contributors/maintainers to keep up with modern Bazel releases and be aware of potential incompatibilities with older versions. But I don't think this testing is all that valuable without releases to represent these kinds of changes.

**Comment by UebelAndre** [comment]
@mfarrugi @damienmg In the issue linked above, a user had an issue due to change that broke compatibility with older versions of Bazel. I think `rules_rust` should have releases and they should indicate breakages or changes in the API but if nothing else I believe the changes in https://github.com/bazelbuild/rules_rust/pull/463 call for a new release to communicate.

**Comment by UebelAndre** [comment]
Also having thought about this a bit. It'd be kinda nice to have a release at https://github.com/bazelbuild/rules_rust/commit/9a8ef691b8e8f682d767189c38339cbee16d0a16 so it's easier to communicate in `cargo-raze` what it's target compatibility version is. Since right now `cargo-raze` is going to potentially generate files that expect newer platform configurations to be added to the repo.

**Comment by UebelAndre** [comment]
Pinging this again. Can these rules please start doing releases? 🙏

**Comment by UebelAndre** [comment]
@hlopko @illicitonion Any chance this effort can be revived? 🙏

**Comment by UebelAndre** [comment]
Pinging this thread again. I think these rules need to adopt a release cadence. It's incredibly taxing for users to have to go digging through commit history to find out what changes they're picking up in any update. A release with a TLDR of the changes would solve for this and is a pattern used by many other rules repositories.

**Comment by hlopko** [comment]
Thanks for pinging.

Not much changed on my end, I still don't have the neither the bandwidth nor the need for regular releases.

If the main problem is figuring out what changed, I'd be supportive to start keeping a CHANGELOG.md file. There I imagine we would mention notable changes and provide links for relevant PRs.

If there is enough demand from the community (and it seems even @UebelAndre alone checks that box :) I have no objections. Logistics of Github releases require repo write access, and I'm happy to grant that to @UebelAndre (their ~ 57 PRs are quite a record anyway :)

What do others think? @mfarrugi @damienmg @illicitonion @dfreese to ping a few?

**Comment by damienmg** [comment]
As I said in a mail before I am already supportive in extending the support team (and preparing my retiring from the team maintainers) and @UebelAndre definitely contributions exceed the level that I expect from a maintainer for rules_rust.

I don't have time to do proper maintenance of this repo, nor do I depend on regular release (I am living at HEAD like a true warrior :p).

**Comment by hlopko** [comment]
Hi all, after chatting with many of you I believe we reached a consensus. I'll go ahead and grant @UebelAndre write access. Andre, feel free to use this issue or create a new one in which we can polish details of releases.

A common theme in many discussions was the lack of clear backwards compatibility policy and unclear ownership (for example I clearly can't be trusted anywhere near cargo-related code yet my approval counts :) I'm currently on vacation (coming back next week), and I plan to start these discussions. Feel free to beat me to it ;)

**Comment by UebelAndre** [comment]
So now that there's functionality for getting a release out there, there are some questions:
1. What are all the things required to actually click the button? 
2. What should go in release notes? 
3. How do we handle the fact that the last release was years ago?
4. What should the new release number be?

I'd love any and all input. I'm really hoping to have a release in the near future.

**Comment by illicitonion** [comment]
> So now that there's functionality for getting a release out there, there are some questions:

🎉 

My own question first:
0. When should we release, and how should we number them?

I think we should release whenever there's a significant non-experimental feature (e.g.: rustfmt aspect, `--error-format` flag), or bugfix to a non-experimental feature (e.g. fixing trailing slashes in build script env vars, fixing link args when interfacing with `cc_library`). i.e. we should release whenever someone may want to pull in a new version. It's fine to batch multiple changes together into a release, but we shouldn't ordinarily delay a release by more than a week because we're waiting for something else - get releases into people's hands early.

I'd suggest we aim for semver starting at `0.1.0` and aim to generally be bumping the patch version (because we try to avoid breaking changes to public APIs), and bump minor if we break public APIs.

I think we should aim to avoid breaking API changes, and they should result in minor version bumps.

I'd also suggest that modifying things in `//cargo/...` to be more true to cargo should not be considered breaking changes, even if they otherwise would be. (e.g. if we find we're setting an env var differently to how cargo would, and we fix that, that's a patch bump, even though it's an observable change in behaviour which may break some compatibility).

I would not consider changes to default or minimum-supported rust of bazel versions to be breaking changes, but we should aim to support the last two LTS releases of Bazel (i.e. 4.x and 5.x), as well as bazel at HEAD, and support at least (completely arbitrarily) 1 year of Rust releases.

> 1. What are all the things required to actually click the button?

a. Ascertain whether the release is breaking, so we can choose the next version number.
b. Ensure CI is green.
c. Make sure release notes are updated (this may be automated).
d. Press the button.

> 2. What should go in release notes?

* A list of any breaking changes
* A list of any noteworthy changes (i.e. anything that a user may be interested in, or which has any risk associated)
* Links to PRs or commits for each of the above
* The minimum and maximum known-good bazel and rust versions

> 3. How do we handle the fact that the last release was years ago?

Ignore them, but start at 0.1.0 (the last release was 0.0.7).

> 4. What should the new release number be?

0.1.0

**Comment by UebelAndre** [comment]
I fully agree with everything @illicitonion said in https://github.com/bazelbuild/rules_rust/issues/415#issuecomment-836539163.

My hope is that releases become a light weight thing and are easy to push out.

For now, I think we should create issues to track all the things we want to describe in a release until that process is automated. I think this issue is fine to act as the release request for `0.1.0`.

@illicitonion @hlopko @dfreese Anything in particular you'd like in the release notes? Other than the obvious "here's how you use this release (`http_archive`) and these are the min supported version".

**Comment by dfreese** [comment]
My initial thoughts were generally:

Only //rust/defs.bzl is considered a part of the versioned public API initially.  The only supported platform is Linux x86-64 currently, with best effort for Windows and Mac or other architectures.

We maintain compatibility with at least all versions of Bazel released within the past 6 months, (in addition to @illicitonion's suggestion of the last 2 LTS releases).

We use semantic versioning, with the following details specific for rules_rust
* Major
  * Compatibility break of public api (i.e. //rust/defs.bzl)
  * Change in the min supported version (potentially could break this into major v. minor Bazel versions, but since bumping Bazel touches everything for users, I wanted to be conservative on this)
* Minor
  * Features added to the public API surface
  * Adding compatibility to a new platform
* Patch
  * Any changes inside of //rust/private that fix or improve behavior, unless they significantly alter platform compatibility or behavior.
  * To optionally communicate any significant changes to code outside of the //rust/defs.bzl

I'd lean for just going with v1.0.0 to give us flexibility for being more clear between major/minor/patch.

I'm leaning this way, because I think it's most important that rules_rust really focus on rust functionality.

This is, in part driven by the my thought that cargo_universe feels to me like a fork of cargo_raze, based on PRs https://github.com/google/cargo-raze/pull/407, https://github.com/google/cargo-raze/pull/412, and https://github.com/google/cargo-raze/pull/415.  Given the work that is going into bootstrapping cargo universe, it seems like it would benefit from being able to semantically version itself independently of rules rust.

I think we'll end up with a much clearer API for rules_rust if we think about them separately.

**Comment by UebelAndre** [comment]
> This is, in part driven by the my thought that cargo_universe feels to me like a fork of cargo_raze, based on PRs [google/cargo-raze#407](https://github.com/google/cargo-raze/pull/407), [google/cargo-raze#412](https://github.com/google/cargo-raze/pull/412), and [google/cargo-raze#415](https://github.com/google/cargo-raze/pull/415). Given the work that is going into bootstrapping cargo universe, it seems like it would benefit from being able to semantically version itself independently of rules rust.

I agree it feels like a fork. Though, I don't want to independently version things in this repo. If we think something should be independently versioned then I question whether or not it belongs in the project. Though that said, I think a tool for managing external dependencies makes sense to be in the rules. I'd vote for keeping `crate_universe` in the project (or in consideration for the release versions once it's moved out of 'experimental'). Maybe that's tangential though. In general I like the idea of having more flexibility in releases.

**Comment by hlopko** [comment]
(cross linking this discussion with https://github.com/bazelbuild/rules_rust/issues/600).

My take on this. I acknowledge that I have a bit of PTSD after heated internal discussions about Bazel backwards compatibility policy, and there are surely areas where I am illogical and biased. Please call them out :) I'll also repeat that I don't plan to use releases, but of course I want to be mindful to our users and I believe some amount of "overhead" to keep the rules backwards compatible or to keep them working with older Bazel versions is warranted.

### Starting point

I prefer not to start with 1.0 right away, I'd like to first gain some experience with the flow, infrastructure, and backwards compatibility. I'd also assume 1.0 marks a milestone, and there are things we want to have in rules rust by then, like mac and windows support. I propose to start with 0.1.0 and leave this question for a separate discussion.

### Release frequency

I believe frequent releases are better. As long as we reasonably can I prefer to release based on features/bug fixes, not on fixed time schedule.

### What is considered backwards compatible

* As @dfreese said, only Linux & x86_64 is currently subject to backwards compatibility policy. Process for moving a best effort platform to a supported platform is consensus based.
* `//rust:defs.bzl` is subject to the backwards compatibility policy (not sure what that is though and I'm OK with discussing that in a separate issue :)
* `//rust/private/...` is **not** subject to the backwards compatibility policy. Content of this package is an implementation detail.
* `//cargo`, `//util`, `//tools`, `//test`, `//examples`, `//bindgen`, `//proto` are by default not subject to the backwards compatibility policy. I'd like to make `//tools` and `//util` backwards compatible after we review what's there and we move implementation details into `//rust/private/tools` or `//rust/private/util`. `//examples` and `//test` stay backwards incompatible. For other packages I'm not qualified to have an opinion :)
* Bug fixes are **not** a breaking change by default. We'll use Common Sense (and we will pull in more maintainers to discuss) if we see a certain bug fix is controversial.

### How to make a backwards compatible change

* Create a GitHub issue (example: https://github.com/bazelbuild/rules_rust/issues/591)
  * Describe the change, motivation for the change, provide migration instructions/tooling.
* Add a build setting if possible that enables the new behavior (or removes the old behavior), mention the issue in error messages.
* Somehow make sure the issue is mentioned in the release notes.
* Give the community time to migrate. I'm curious what folks think is long enough. 3 months since the first release mentioning the issue? 6 months? More?

### Bazel compatibility

How strongly do you feel about supporting last 2 Bazel LTS releases? Last LTS and Bazel at HEAD sound good to me. I'd also be fine with having a limited window where we support 2 LTSs when Bazel releases a new LTS.

### Release notes

What @illicitonion said. I'm not sure how other projects maintain their changelogs. Maybe we can get inspiration from https://github.com/rust-analyzer/rust-analyzer/blob/e529f733550e9d7bd0d9558a455bbbb5906b917c/docs/dev/README.md#release-process. It seems to be scanning PR descriptions.

### Version numbers

I think what @illicitonion and @dfreese said makes perfect sense and is to a large extent trivially mergeable. Whether to have different version number for crate universe and rules rust I don't know. I'd hope that having them be released and versioned together will not add too much noise for users.

**Comment by illicitonion** [comment]
> ### How to make a backwards compatible change
> * Create a GitHub issue (example: #591)
>   
>   * Describe the change, motivation for the change, provide migration instructions/tooling.
> * Add a build setting if possible that enables the new behavior (or removes the old behavior), mention the issue in error messages.
> * Somehow make sure the issue is mentioned in the release notes.
> * Give the community time to migrate. I'm curious what folks think is long enough. 3 months since the first release mentioning the issue? 6 months? More?

3-ish months sounds fine.

I'm also assuming that we're unlikely to back-port fixes to previous releases (i.e. in practice we're always moving forwards with the latest release - after `x.y.0` is cut, we're unlikely to release a new `x.(y-1).z`).

> ### Bazel compatibility
> How strongly do you feel about supporting last 2 Bazel LTS releases? Last LTS and Bazel at HEAD sound good to me. I'd also be fine with having a limited window where we support 2 LTSs when Bazel releases a new LTS.

I think that overlap window is important. I think I'm fine with supporting just one Bazel LTS release _after we've supported both_ for say 3 months of actual use (i.e. after the new LTS is released and actually use by someone), but as a user, not having to futz with changing rules and potentially changing rule APIs when bumping Bazel versions is really attractive.

---

I also think that in practice it's important to be pragmatic. If Bazel 5 decides that providers are going to be replaced with a completely new concept, then sure, we may not support two LTSs concurrently. If we make a breaking change and after 2 months we really need to remove the old API, or we really need to keep the old API for 6 months, we should be fine with that. But it's good to have a shared set of guidelines we strive towards :)

**Comment by dfreese** [comment]
> > This is, in part driven by the my thought that cargo_universe feels to me like a fork of cargo_raze, based on PRs [google/cargo-raze#407](https://github.com/google/cargo-raze/pull/407), [google/cargo-raze#412](https://github.com/google/cargo-raze/pull/412), and [google/cargo-raze#415](https://github.com/google/cargo-raze/pull/415). Given the work that is going into bootstrapping cargo universe, it seems like it would benefit from being able to semantically version itself independently of rules rust.
> 
> I agree it feels like a fork. Though, I don't want to independently version things in this repo. If we think something should be independently versioned then I question whether or not it belongs in the project. Though that said, I think a tool for managing external dependencies makes sense to be in the rules. I'd vote for keeping `crate_universe` in the project (or in consideration for the release versions once it's moved out of 'experimental'). Maybe that's tangential though. In general I like the idea of having more flexibility in releases.

Based on https://github.com/bazelbuild/rules_rust/pull/743, I think it's time to start looking for a home outside of rules_rust for cargo_universe.  Open source projects work well when they do one thing, and when they do one thing well.  Forking cargo-raze into this project brings a significant maintenance burden both for this project, and for anyone attempting to make sure significant fixes or changes make it into both.  https://github.com/google/cargo-raze/pull/282 is a clear example of a change that would become nearly impossible even with a small divergence in the code bases.

**Comment by UebelAndre** [comment]
> > > This is, in part driven by the my thought that cargo_universe feels to me like a fork of cargo_raze, based on PRs [google/cargo-raze#407](https://github.com/google/cargo-raze/pull/407), [google/cargo-raze#412](https://github.com/google/cargo-raze/pull/412), and [google/cargo-raze#415](https://github.com/google/cargo-raze/pull/415). Given the work that is going into bootstrapping cargo universe, it seems like it would benefit from being able to semantically version itself independently of rules rust.
> 
> > 
> 
> > I agree it feels like a fork. Though, I don't want to independently version things in this repo. If we think something should be independently versioned then I question whether or not it belongs in the project. Though that said, I think a tool for managing external dependencies makes sense to be in the rules. I'd vote for keeping `crate_universe` in the project (or in consideration for the release versions once it's moved out of 'experimental'). Maybe that's tangential though. In general I like the idea of having more flexibility in releases.
> 
> 
> 
> Based on https://github.com/bazelbuild/rules_rust/pull/743, I think it's time to start looking for a home outside of rules_rust for cargo_universe.  Open source projects work well when they do one thing, and when they do one thing well.  Forking cargo-raze into this project brings a significant maintenance burden both for this project, and for anyone attempting to make sure significant fixes or changes make it into both.  https://github.com/google/cargo-raze/pull/282 is a clear example of a change that would become nearly impossible even with a small divergence in the code bases.

I think this is a separate conversation from releases. Maybe you can start the conversation on that PR or in a new issue?

**Comment by illicitonion** [comment]
> > > This is, in part driven by the my thought that cargo_universe feels to me like a fork of cargo_raze, based on PRs [google/cargo-raze#407](https://github.com/google/cargo-raze/pull/407), [google/cargo-raze#412](https://github.com/google/cargo-raze/pull/412), and [google/cargo-raze#415](https://github.com/google/cargo-raze/pull/415). Given the work that is going into bootstrapping cargo universe, it seems like it would benefit from being able to semantically version itself independently of rules rust.
> > 
> > 
> > I agree it feels like a fork. Though, I don't want to independently version things in this repo. If we think something should be independently versioned then I question whether or not it belongs in the project. Though that said, I think a tool for managing external dependencies makes sense to be in the rules. I'd vote for keeping `crate_universe` in the project (or in consideration for the release versions once it's moved out of 'experimental'). Maybe that's tangential though. In general I like the idea of having more flexibility in releases.
> 
> Based on #743, I think it's time to start looking for a home outside of rules_rust for cargo_universe. Open source projects work well when they do one thing, and when they do one thing well. Forking cargo-raze into this project brings a significant maintenance burden both for this project, and for anyone attempting to make sure significant fixes or changes make it into both. [google/cargo-raze#282](https://github.com/google/cargo-raze/pull/282) is a clear example of a change that would become nearly impossible even with a small divergence in the code bases.

I'm not convinced #743 is a good idea, largely for the reasons you mention - I think as much of the "understanding Cargo" pieces should live not-in-`rules_rust`, and it would be a shame if we ended up with multiple competing rather than collaborating efforts to implement this piece.

I've tried pretty hard to make sure that the changes I've put together in this area work well for cargo-raze _and_ consumers of it. For example, https://github.com/google/cargo-raze/pull/412 which fixes a 3-year-old feature request in cargo-raze while _also_ allowing things built on cargo-raze to benefit in the same way.

> cargo_universe feels to me like a fork of cargo_raze, based on PRs [google/cargo-raze#407](https://github.com/google/cargo-raze/pull/407), [google/cargo-raze#412](https://github.com/google/cargo-raze/pull/412), and [google/cargo-raze#415](https://github.com/google/cargo-raze/pull/415)

I'd love to change that feeling, and instead view cargo-raze as two separate components - a "cargo metadata on steroids" (which can be re-used - ideally as a library, but if not, with a JSON output mode or similar) and a static up-front BUILD-file-generator.

On those three PRs specifically, I agree that google/cargo-raze#407 is problematic and probably shouldn't merge. I'm not sure why google/cargo-raze#412 is problematic (as described above). google/cargo-raze#415 feels to me like precisely the opposite of a fork - it's an effort to share logic to _avoid_ a fork.

If the messaging here is that cargo-raze as a project _isn't_ interested in being the repo that owns this logic, then probably a fork of that logic (probably in a separate repo, and not in `rules_rust`) makes sense, but I'd really rather we all work together if we can come to an agreement about what that looks like - I think we all have very similar aims, and a lot of knowledge and experience which can support all of the relevant efforts here!

If `cargo-raze` itself isn't the right vehicle for this, how would folks feel about pulling out a separate crate containing basically the metadata fetching and build planning stages of `cargo-raze` which both projects could cleanly depend on? `cargo-raze` is already architected nicely into a pipeline of stages, and extracting some of them to be re-usable could perhaps help to make a boundary more clear.

---

The big value that I see of `crate_universe` being part of `rules_rust` is a solid compatibility story. If we're talking about splitting those apart, I'd want to make it trivial to answer the questions "I'm using `rules_rust` at release `X`, what `crate_universe` should I be using?" and "I'm using `crate_universe` at release `Y`, what `rules_rust` should I be using?". This is trivially satisfied by them being in the same repo and same releases. I think the big question to answer is: **What's the compatibility story around the `cargo_build_script` rule (i.e. the contents of `@rules_rust//cargo/...`)?**

If we're saying that we long-term want to keep the generation of `cargo_build_script` externalised, we need to include `cargo_build_script` in our stability guarantees.

It's not viable to say that external tools should be generating a moving target, and we already see difficulties with `cargo-raze` with this today, where when APIs change in `rules_rust` it's unclear what versions of `cargo-raze` work with them (and vice-versa - when `cargo-raze` starts supporting new functionality in `rules_rust` it's unclear what versions of `rules_rust` it works with).

And if we _do_ offer some kind of stability guarantee, **how do we make sure we enforce that guarantee in CI?** The convenience of having `crate_universe` inside the repo, with its executable examples, is that we proactively find out if we break an API. This plugs the gap of an API testing piece very conveniently, but it's not the only way - we could hand-craft similar rules, we could pull in some examples using a fixed version of `cargo-raze` or `crate_universe` (or both), or there are potentially other approaches. It's not clear to me which approach has the lowest maintenance burden here.

Another option could be to remove `cargo_build_script` from the repo completely, but then we would need to standardise something like the `BuildInfo` provider (which is intricately tied to the `cargo_build_script` implementation details, and I suspect doesn't really make much sense on its own as a public API...)

**Comment by UebelAndre** [comment]
# 
> Forking cargo-raze into this project brings a significant maintenance burden both for this project, and for anyone attempting to make sure significant fixes or changes make it into both. [google/cargo-raze#282](https://github.com/google/cargo-raze/pull/282) is a clear example of a change that would become nearly impossible even with a small divergence in the code bases.

I don't think this would realistically happen. I think folks will either use `cargo-raze` or they will use `crate_universe` and I think the usability of reviewing one or two files for a change which updates dependencies is worth adopting `crate_universe` over `cargo-raze`. The vendoring of these build files is a massive distraction in PRs and makes it far more difficult to review than it should.

> Based on #743, I think it's time to start looking for a home outside of rules_rust for cargo_universe. Open source projects work well when they do one thing, and when they do one thing well.

I disagree that `crate_universe` belongs in a separate project. A large part of the Rust ecoecosystem is code sharing via crates. The rules should offer a way to interact with external dependencies in a reasonable fashion.

#

> I'm not convinced #743 is a good idea, largely for the reasons you mention - I think as much of the "understanding Cargo" pieces should live not-in-`rules_rust`, and it would be a shame if we ended up with multiple competing rather than collaborating efforts to implement this piece.

The intent of that PR is to isolate the "metadata fetching" and planning components that currently live in `cargo-raze` and put them in a state where progress `crate_universe` is no longer being negatively impacted. Having to make PRs into `cargo-raze` in order to get the functionality in `crate_universe` is harmful to productivity and progress for a couple reasons:

1. There's next to no review support
In order to get anything merged, it seems you have to borderline harass someone to finally look at your change. It's already known that maintainer availability is bad in `rules_rust` where there are ~5 maintainers but it's almost impossible to get anything merged in a project (`cargo-raze`) where there's really only 1-2. There is also the added affect where the infrequent reviews lead to a loss of context on changes happening in the repo. I feel pretty comfortable with `cargo-raze` having either updated or refactored nearly every module but trying to review https://github.com/google/cargo-raze/pull/282 and integrate them into https://github.com/bazelbuild/rules_rust/pull/743 was a big challenge given how long it'd been since I looked at `cargo-raze`. Unless there's active reviewers and a clear direction, the project seems like it's on it's way to being dead (it certainly felt that it was at one point).

2. Some changes don't make sense for `cargo-raze` but would for `crate_universe`.
https://github.com/google/cargo-raze/pull/407 is something that honestly should be a deletion of that functionality entirely. Not having the dependencies required by that feature meant `crate_universe` was buildable without requiring some pretty large and hard to build dependencies for functionality that makes no sense given what the use case was. Over time, there have been a number of changes opened to `cargo-raze` which did not make sense for the tool as it was but does make sense for either a library for getting metadata or `crate_universe` itself (in no particular order):
- https://github.com/google/cargo-raze/pull/311  No functional change for `cargo-raze` only for the `crate_universe` use case, which I only learned about later.
- https://github.com/google/cargo-raze/pull/222 An approach to gathering additional metadata but outside of the metadata module. I didn't understand the rational here until I heard about `crate_universe`
- https://github.com/google/cargo-raze/pull/415 a desire to make structs that house the updated metadata easily and consistently writable to files. This is not functionality `cargo-raze` has a use for.

I'm confident there will be more over time as `crate_universe` sees more and more use.

> If `cargo-raze` itself isn't the right vehicle for this, how would folks feel about pulling out a separate crate containing basically the metadata fetching and build planning stages of `cargo-raze` which both projects could cleanly depend on? `cargo-raze` is already architected nicely into a pipeline of stages, and extracting some of them to be re-usable could perhaps help to make a boundary more clear.

Separating this functionality out would solve for my second issue above but not the first. I'm tired of `cargo-raze` being a bottle neck in my workflow and having to spend time maintaining a fork while I wait months for a reviewer to come around and begin to provide feedback.

`cargo-raze` is an awesome tool and in my opinion has been critical in getting the core rules to where they are today (which is a great spot). But unfortunately no one is continuing to grow the project and has entered a sort of maintenance mode. It may not currently be the dream approach to managing external dependencies, but it's more than sufficient in unblocking this functionality and putting us in a position where we can incrementally get there. It should be in all our best interest to put the code in a state where we can work together to ensure it's continued growth into the thing we want for these rules. Given that there was already a discussion acknowledging the lack of support in `cargo-raze` https://github.com/google/cargo-raze/discussions/314, I feel the smartest and most effective change is to move the metadata and planning logic into `rules_rust` and use our collective knowledge here to create something that is focused and intended to be used as a library to support projects like `crate_universe` and `cargo-raze`.

**Comment by dfreese** [comment]
Honestly forgot that cargo_build_script was outside of //rust.  While I'd generally appreciate a brighter line between rustc and cargo, rust_binary/library has enough cargo-isms already that I think I'd be alright supporting that rule as a part of the public API, assuming that's a sticking point here.

Hadn't looked at https://github.com/google/cargo-raze/pull/412, so I don't have a strong opinion there.  Was generally fine with https://github.com/google/cargo-raze/pull/415, but @UebelAndre had said he wanted to wait on that for https://github.com/google/cargo-raze/pull/282.

The main issue I had was with https://github.com/google/cargo-raze/pull/407.  I was and am open to having a separate crate that exposes the metadata fetching and build planning:
> Got it. So the typical approach here would be to pull things out into a -core or -base crate. ... Is there was a way for this feature break to happen strictly at a module level? That will probably help with the discussion here.

However, there wasn't any further discussion specifically on the module question.

**Comment by UebelAndre** [comment]
>  I was and am open to having a separate crate that exposes the metadata fetching and build planning:

My concern in starting an effort like this is you are really the only person who looks at `cargo-raze` and only do so every once in a while. Prior attempts to get anyone's attention there generally go unresolved or take months. I don't really care where this code lives, but I do care about the turn-around time for changes and the ability for a larger group to collaborate and contribute to it. Again, I'm tired of `cargo-raze` feeling like a blocker to progress. Any suggestions on how to address that?

**Comment by illicitonion** [comment]
It sounds like there _may_ be some agreement about pulling out a shared crate for metadata + planning, and having both `cargo-raze` and `crate_universe` depend on that? (What I imagine that would look like for #407 would be that potentially there would be a few odd warts in the API of that crate so that `cargo-raze` could pass in some kind of callback to a few methods to add the binary_deps information, and `crate_universe` could pass in a no-op callback, but that's workable)...

Perhaps if we did so, that crate could have a wider ownership, so that `cargo-raze` doesn't fundamentally change its offering, but some of the implementation details would end up with a better-specified API (which would help make them more intelligible) which could be contributed to, and reviewed, by a wider pool of people?

**Comment by UebelAndre** [comment]
re: https://github.com/bazelbuild/rules_rust/issues/415#issuecomment-843407093 
I think pulling a crate for metadata + planning is a fantastic idea and agree that crate should have wider ownership.

**Comment by hlopko** [comment]
Do I understand correctly that `crate_universe` topic is resolved?

Any other comments on backwards compatibility? If not I can write up the policy and upload a PR for you all to review.

**Comment by UebelAndre** [comment]
> Do I understand correctly that `crate_universe` topic is resolved?

There has yet to be consensus from @dfreese and there hasn't been a discussion on where that crate should live. My vote is here in `rules_rust`. That may also be relevant to release conversations but I do think that's something that can happen as an aside to you getting a PR open.

**Comment by UebelAndre** [comment]
Any updates here?

**Comment by UebelAndre** [comment]
@hlopko hey, another ping to see if I could get you to write up the policy so we could get a release out 🙏

**Comment by sayrer** [comment]
This issue is getting fairly difficult to navigate as a consumer of both `rules_rust` and `cargo-raze`.

I did this PR: https://github.com/google/cargo-raze/pull/425

I don't have any strong opinions about the resolution here, but this issue has been open for almost a year.

**Comment by sayrer** [comment]
> It sounds like there _may_ be some agreement about pulling out a shared crate for metadata + planning, and having both `cargo-raze` and `crate_universe` depend on that?

Having done enough work on an internal fork of `cargo-raze`, which adds local crates and tests, I think this idea makes sense. The rendering code that `cargo-raze` supplies is a perfectly reasonable default, but it is not very easy to customize.

**Comment by UebelAndre** [comment]
So now that https://github.com/bazelbuild/rules_rust/pull/832 is done, what's in the way of making a release? 😄

**Comment by hlopko** [comment]
Well, you may know better than me :) I guess we can go ahead and implement the release pipeline?

**Comment by illicitonion** [comment]
I _think_ we reached rough consensus on this - do we have blockers before releasing our first stable version?

@hlopko @UebelAndre @dfreese?

**Comment by UebelAndre** [comment]
In general I'm on board with doing releases though I'd like to *not* include `crate_universe` binaries in the first one since I don't think the pattern there are in a good state.

**Comment by hlopko** [comment]
**Do you mean 0.X or 1.0?**

I'd vote for creating the automation (actual releases, changelog management?) to create releases, create a bunch of 0.X releases, and then when bugs are discovered and fixed we can create 1.0. 

**What is in scope for the release?**

`//rust` obviously. What about `//proto` and `//bindgen`? I presume `//test` and `//examples` are out. `//tools` and `//utils` are in.

**Anything we want to do before 1.0?**

There are some pet peeves of mine to which I didn't get to yet, like https://github.com/bazelbuild/rules_rust/issues/843, https://github.com/bazelbuild/rules_rust/issues/637, https://github.com/bazelbuild/rules_rust/issues/622, https://github.com/bazelbuild/rules_rust/issues/601. I think I should start accepting that I may not fix all of them before 1.0 :)



What do others think?

**Comment by illicitonion** [comment]
> **Do you mean 0.X or 1.0?**
> 
> I'd vote for creating the automation (actual releases, changelog management?) to create releases, create a bunch of 0.X releases, and then when bugs are discovered and fixed we can create 1.0.

+1

> **What is in scope for the release?**
> 
> `//rust` obviously. What about `//proto` and `//bindgen`? I presume `//test` and `//examples` are out. `//tools` and `//utils` are in.

Also `//cargo` but not `//crate_universe`.

> **Anything we want to do before 1.0?**
> 
> There are some pet peeves of mine to which I didn't get to yet, like #843, #637, #622, #601. I think I should start accepting that I may not fix all of them before 1.0 :)
> 
> What do others think?

I can be swayed to block on any of those _as long as they're going to be actively worked on_. I don't think any of them are important enough to set them as indefinite blockers except maybe #601.

**Comment by UebelAndre** [comment]
Finally getting back to this. Does anyone have an objection to releasing `rules_rust v0.1.0` by the end of the week?

**Comment by hlopko** [comment]
I'd say go ahead :)

**Comment by UebelAndre** [comment]
Started a release: https://github.com/bazelbuild/rules_rust/actions/runs/2004339369

**Comment by UebelAndre** [comment]
oh no! it failed 😞

**Comment by UebelAndre** [comment]
Going to try to recreate it.

**Comment by UebelAndre** [comment]
created https://github.com/softprops/action-gh-release/issues/221 to track the failure.

**Comment by UebelAndre** [comment]
Round 2 was a success! https://github.com/bazelbuild/rules_rust/actions/runs/2004511376

**Comment by UebelAndre** [comment]
https://github.com/bazelbuild/rules_rust/releases/tag/0.1.0 looks pretty epic

**Comment by UebelAndre** [comment]
Trying now to make sure the release is properly uploaded to the Google mirror.

**Comment by UebelAndre** [comment]
https://github.com/bazelbuild/continuous-integration/issues/1347 tracks potentially automating this step

**Comment by UebelAndre** [comment]
With `0.1.0` out, and the release actions looking good. This ticket is now complete! Thanks everyone who contributed to this!
